teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,478 @@
+- Based on SLSA L2 requirements, the provenance must provide the location of source code in version control and the provenance MUST identify the output artifact via at least one cryptographic hash.
+- If a TaskRun is within a PipelineRun, the TaskRun’s provenance can be produced between the time that when one TaskRun is finished
+
+
+## Proposal
+
+Currently we are using type-hinting to retrieve needed provenance info from TaskRun’s params and results. The proposal aims to leverage the new features of structured params and results described in [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support better structured provenance retrieval for inputs and artifacts to make the current type-hinting easier. This proposal is focused on the use cases for generation in-toto attestations, as other types of attestations don’t need the TaskRun params and results.
+
+The workflow for the targeted results and params in Chains is like the following:
+```
+(TaskRun params & results) —-> (Scanned by Chains controller, targets are collected) —> (Targets are parsed and validated and matched against with defined result and param structures)  —-> (Chains controller sign the signable targets and put into intoto attestations)
+```
+As mentioned above, the signable targets can be split into inputs (sources) and outputs (artifacts). Instead of scatteredly collecting provenance pieces from TaskRun, we collect inputs/ provenance from params, and generate outputs provenance from TaskRun results. The signable targets being found in params will be the inputs, and those found in TaskRun results will be the outputs.
+
+Tekton Chains will follow [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support the sets of json schemas and described in the documents. In this proposal, the expected formats of params and results are also defined.
+
+
+### Notes and Caveats
+
+When these well defined structures grow as richer provenance are pursued, the size of the structure can eventually grow beyong the limit of the container termination message. [TEP 0086](https://github.com/tektoncd/community/pull/521) is addressing this issue.
+
+
+## Design Details
+
+### Signable Targets
+
+All signable targets should be provided in a structure that is well defined in the proposal. If any field is missed, the signable targets will be skipped and an error will be thrown.
+
+#### Version Control Sources
+
+Based on the use cases and data storage within provenance described above, it makes sense to have the source object to be the following because all version control sources will be grinded down into the requirements in the Predicate’s materials:
```

Comment:
Just to be explicit here, this isn't a new type we're inventing here right? What we're really trying to do is standardize on SLSA's `materials` field and defintion?

Reply:
Yes. They are not invented from here - these structures are similar because they follow the standards of `materials` in SLSA

Reply:
Since theses are all just variants of the same message schema, perhaps it would make more sense to just define 1 result type and list the various usages (e.g. git, oci, python, ...) under an examples section?